### DISCUSSION TITLE
Client ID, Client Host, and Client IP Address protocol mappers

### DISCUSSION THREAD

**Discussion by jesseschmieg** [keycloak/keycloak#13602] [Q&A]
By default, a client with service accounts enabled automatically has created protocol mappers to map token claims for `clientId`, `clientHost` and `clientAddress`. Even after deleting, there is logic to recreate these mappers upon token request if the Client ID mapper is not present. I'm curious about 1) if these claims serve some purpose beyond consumers of the tokens (for example, logging information when the tokens are used with Keycloak APIs) and 2) is there any way to avoid creating these mappers by default?

Thank you - any help is appreciated.

**Comment by oemergenc** [comment]
Hi, 

we are also stumbled over this. Can someone clarify why these mappers are created and are we safe to delete them? Up to now, we are not sure where and what exactly is the result of those mappers? Some insights would be nice. Thanks.

**Comment by alefcarlos** [comment]
Same here

**Comment by cgeorgilakis** [comment]
Client Id mapper,adding client_id claim, is based on OAuth2 specifications about client_id claim:

- [JSON Web Token (JWT) Profile for OAuth 2.0 Access Tokens](https://datatracker.ietf.org/doc/html/rfc9068#section-2.2)

- [OAuth 2.0 Token Introspection](https://www.rfc-editor.org/rfc/rfc7662#section-2.2)

- [OAuth 2.0 Token Exchange](https://datatracker.ietf.org/doc/html/rfc8693#section-4.3)

If I am not wrong, this claim is related also to findind client user in token introspection and user info endpoint.
I believe that other two mappers are useless.
